### PR TITLE
HPI with `_SearchSpaceTransform`

### PR DESCRIPTION
**Pull request by hvy**
## Motivation

Refactoring of HPI.

## Description of the changes

Somewhat simplifies implementation details of HPI using `_SearchSpaceTransform`. See https://github.com/optuna/optuna/pull/1987.

## Note

Depends on https://github.com/optuna/optuna/pull/1987. Probably easier to grasp by checking only the relevant commits.

### CODE REVIEW

**Comment by HideakiImamura** [comment]
#1987 has been merged. Could you merge the master branch?

**Comment by hvy** [comment]
Thanks for the heads up, merge with master. PTAL!

File: optuna/_transform.py [hunk 1]
```diff
@@ -46,6 +46,10 @@ class _SearchSpaceTransform:
     Attributes:
         bounds:
             Constructed bounds from the given search space.
+        column_to_encoded_columns:
+            Constructed mapping from original parameter column index to encoded column indices.
+        encoded_column_to_column:
+            Constructed mapping from encoded column index to original parameter column index.
```
Comment (hvy) [inline optuna/_transform.py:52]: I can create a separate PR for the transform changes if you'd like. This PR might give some context.

**Review by HideakiImamura — APPROVED** [review]
Thanks for the PR! This simplifies the implementation of HPI modules and does not change the logic. LGTM!

**Comment by HideakiImamura** [comment]
@hvy Could you merge the master branch to resolve the conflicts?

@sile Could you review this PR if you have time?

**Comment by sile** [comment]
Sure!

**Comment by hvy** [comment]
Thanks for the heads up, totally missed the conflicts. They were by the way caused by the `target` argument so I had to make some non-trivial changes during merge https://github.com/optuna/optuna/pull/1988/commits/446f584c887bf2deb2b233acc0a736edd78b216d

**Review by sile — APPROVED** [review]
LGTM!

**Pull request merged**